cranelift/fuzzgen/src/function_generator.rs

```diff
@@ -39,7 +40,19 @@ fn insert_opcode(
     // For pretty much every instruction the control type is the return type
     // except for Iconcat and Isplit which are *special* and the control type
     // is the input type.
-    let ctrl_type = if opcode == Opcode::Iconcat || opcode == Opcode::Isplit {
+    let ctrl_type = if [
+        Opcode::Iconcat,
+        Opcode::Isplit,
+        Opcode::UwidenHigh,
+        Opcode::SwidenHigh,
+        Opcode::UwidenLow,
+        Opcode::SwidenLow,
+        Opcode::Snarrow,
+        Opcode::Unarrow,
+        Opcode::Uunarrow,
+    ]
+    .contains(&opcode)
+    {
```

Comment:
I think [OpcodeConstraints::requires_typevar_operand](https://docs.rs/cranelift-codegen/latest/cranelift_codegen/ir/instructions/struct.OpcodeConstraints.html#method.requires_typevar_operand) is most of what we need to avoid having to update this table:

```suggestion
    let ctrl_type = if opcode.constraints().requires_typevar_operand() {
```

Also please update the comment above.

The rest of what we need is figuring out which operand is the typevar operand. It is _almost_ always operand 0, but it depends on the instruction format. For `Ternary` it's operand 1, and for `TernaryImm8` it's operand 2. (See `cranelift/codegen/meta/src/shared/formats.rs` and look for calls to `InstructionFormatBuilder::typevar_operand`.)

Since `insert_opcode` is used for the `Ternary` format, this technically would look at the wrong typevar operand. Fortunately, all current `Ternary`-format opcodes have `requires_typevar_operand=false`, so we don't actually have to fix this today.

As future work, I think we should provide a method to get the index of the typevar operand. I think it should be on `InstructionFormat`. There's an `InstructionData::typevar_operand` method today that gets the `Value` from the appropriate operand of an instruction, but here we need the operand index. (We could replace the existing `typevar_operand` method by calling this new one and indexing into `InstructionData::arguments`, I think.)

Reply:
> I think [OpcodeConstraints::requires_typevar_operand](https://docs.rs/cranelift-codegen/latest/cranelift_codegen/ir/instructions/struct.OpcodeConstraints.html#method.requires_typevar_operand) is most of what we need to avoid having to update this table:

Oh that's really nice! I didn't know that existed.

> As future work, I think we should provide a method to get the index of the typevar operand. I think it should be on InstructionFormat. There's an InstructionData::typevar_operand method today that gets the Value from the appropriate operand of an instruction, but here we need the operand index. (We could replace the existing typevar_operand method by calling this new one and indexing into InstructionData::arguments, I think.)

I was thinking about this in a slightly different way. With the new fancy constraint based generation we always know the correct `ctrl_type`, so we could update the inserter signature to also receive that. There are a bunch of inserters that currently do something like `rets.first().unwrap()` which we can cleanup.

Either way, this can definitely be improved!